provider/kubernetes.go

```diff
@@ -193,28 +205,42 @@ func (provider *Kubernetes) loadIngresses(k8sClient k8s.Client) (*types.Configur
 						if port.Port == 443 {
 							protocol = "https"
 						}
-						endpoints, exists, err := k8sClient.GetEndpoints(service.ObjectMeta.Namespace, service.ObjectMeta.Name)
-						if err != nil || !exists {
-							log.Errorf("Error retrieving endpoints %s/%s: %v", service.ObjectMeta.Namespace, service.ObjectMeta.Name, err)
-							continue
-						}
-						if len(endpoints.Subsets) == 0 {
-							log.Warnf("Endpoints not found for %s/%s, falling back to Service ClusterIP", service.ObjectMeta.Namespace, service.ObjectMeta.Name)
-							templateObjects.Backends[r.Host+pa.Path].Servers[string(service.UID)] = types.Server{
-								URL:    protocol + "://" + service.Spec.ClusterIP + ":" + strconv.Itoa(int(port.Port)),
+						if service.Spec.Type == "ExternalName" {
+							url := protocol + "://" + service.Spec.ExternalName
+							name := url
+
+							templateObjects.Backends[r.Host+pa.Path].Servers[name] = types.Server{
+								URL:    url,
 								Weight: 1,
 							}
 						} else {
-							for _, subset := range endpoints.Subsets {
-								for _, address := range subset.Addresses {
-									url := protocol + "://" + address.IP + ":" + strconv.Itoa(endpointPortNumber(port, subset.Ports))
-									name := url
-									if address.TargetRef != nil && address.TargetRef.Name != "" {
-										name = address.TargetRef.Name
-									}
-									templateObjects.Backends[r.Host+pa.Path].Servers[name] = types.Server{
-										URL:    url,
-										Weight: 1,
+							endpoints, exists, err := k8sClient.GetEndpoints(service.ObjectMeta.Namespace, service.ObjectMeta.Name)
+							if err != nil {
+								log.Errorf("Error while retrieving endpoints from k8s API %s/%s: %v", service.ObjectMeta.Namespace, service.ObjectMeta.Name, err)
+								continue
+							} else if !exists {
```

Comment:
Sure! I'm basically proposing to replace (reducing/simplifying non-relevant parts):

```go
endpoints, exists, err := k8sClient.GetEndpoints(ns, name)
if err != nil {
  log.Errorf("error: %s", err)
  continue
} else if !exists {
  ...
}
```

by

```go
endpoints, exists, err := k8sClient.GetEndpoints(ns, name)
if err != nil {
  log.Errorf("error: %s", err)
  continue
}

if !exists {
  ...
}
```

(Note the difference is one less `else` and an additional line break for readability reasons.)

The logic behind this is as following:

1. If we run into the `err != nil` case, we'll log an error message and `continue` to the next loop iteration. Thus, we will never reach the statement checking for `!exists`.
2. If we run into the `err == nil` case, we want to check for `!exists`. In that sense, it's not necessary to use an `else if` statement because of the previous point: We have already "protected" the `err != nil` case from doing any additional, unnecessary checks (unnecessary because in the case where an error occurred, no further actions should be taken) by a `continue` statement. So overall, a simple `if` suffices.

My rationale here is that it's usually beneficial to turn unnecessary `else if` into more simple `if` statements in order to minimize the cognitive load. (No `else` case means I won't have to bother thinking "what _other_ case are we in now?").

Another reason is that Go programs usually follow a pattern of

1. Calling a function
2. Checking for error
3. in case of an error: return the error
4. Otherwise, proceed to the next function call

The piece of code at hand is a variation of this pattern where we replace 3. by a log and a `continue` statement.

Hope this helps!